### ISSUE TITLE
Error Code 108 with Vive Pro + Ubuntu Bionic Beaver

### ISSUE THREAD

**Issue by georgewsinger** [ValveSoftware/SteamVR-for-Linux#131]
#### Your system information
* Steam client version (build number or date): "Steam package versions: 1536436120"
* Distribution (e.g. Ubuntu): Ubuntu Bionic Beaver 18.04.1 LTS
* Graphics driver version (run nvidia-settings): 396.54
* Gist for SteamVR System Information: [Here](https://gist.github.com/georgewsinger/6618b1bd0965eb34365c4990ad079c3a).
* Opted into Steam client beta?: No (AFAIK)
* Opted into SteamVR beta?: No (but error persists in Beta).
* Have you checked for system updates?: Yes

#### Please describe your issue in as much detail as possible:

![](https://i.imgur.com/9tmwQYJ.png)

As shown above, I am getting an error 108. Some relevant snippets from my logs (see gist above) are as follows:

#### Steps for reproducing this issue:
1. Install a fresh Ubuntu Bionic 18.04 from scratch (with HMD Vive Pro plugged in).
2. Update graphics drivers to nvidia 396.
3. Install steam and SteamVR.
4. Update udev rules per the README.
5. Launch SteamVR.
6. Receive 108 Error.

**Comment by MuraliSridhar** [comment]
same issue.. any help ?

**Comment by georgewsinger** [comment]
@MuraliSridhar I just realized this could be an issue with the Vive Pro + Linux combo. Has even a single person confirmed that the Vive Pro works on Linux?

**Comment by DiegoAG** [comment]
#### Your system information
* Steam client version (build number or date): 1537398481
* Distribution (e.g. Ubuntu): Ubuntu Bionic Beaver 18.04.1 LTS
* Graphics driver version (run nvidia-settings): amdgpu-pro-18.30-641594
* Gist for SteamVR System Information: 
* Opted into Steam client beta?: [Yes/No] NO (AFAIK)
* Opted into SteamVR beta?: [Yes/No] NO (AFAIK) but menu says its beta
* Have you checked for system updates?: [Yes/No] YES

#### Please describe your issue in as much detail as possible:

![captura de pantalla de 2018-09-21 19-31-31](https://user-images.githubusercontent.com/5911377/45896940-f7ed2e00-bdd5-11e8-8892-fd1df69eee7a.png)
As shown above, I expected to start SteamVR with Vive Pro + Vega 64 but I am getting an error 108. Some relevant snippets from my logs (see gist ) are as follows:
https://gist.github.com/DiegoAG/343f34f0bfa0a977c57ef4da19ea5718

#### Steps for reproducing this issue:
1. Install a fresh Ubuntu Bionic 18.04 from scratch.
2. Update graphics drivers to amd latest nowadays amdgpu-pro-18.30-641594
3. Install steam and SteamVR.
4. Update udev rules per the README.
5. Plugin Htc Vive Pro 
6. Launch SteamVR.
7. Receive 108 Error.

**Comment by georgewsinger** [comment]
@DiegoAG Since Diego is getting the error on AMD, the only constants left are Ubuntu Bionic 18.04 or the Vive Pro.

**Comment by lboklin** [comment]
Considering my Vive works on Kubuntu 18.04 I'd say it boils down to the Vive Pro

**Comment by lboklin** [comment]
I don't think you're supposed to run Steam as root.. that sounds like something that might apply to Windows, not Linux.

**Comment by kisak-valve** [comment]
Hello @DiegoAG, Steam on Linux does not benefit in any way from running as root, that's only a needless security risk.

If you're feeling adventurous, you could give my [PPA](https://launchpad.net/~kisak/+archive/ubuntu/steamvr) a try, but I do not have a Vive Pro to test locally.

**Comment by DiegoAG** [comment]
> Hello @DiegoAG, Steam on Linux does not benefit in any way from running as root, that's only a needless security risk.
> 
> If you're feeling adventurous, you could give my [PPA](https://launchpad.net/%7Ekisak/+archive/ubuntu/steamvr) a try, but I do not have a Vive Pro to test locally.

I was a bit surprised (what on earth!!!) about the execution as an administrator . I will try Kisak PPA in 12h and let you know.
Thanks in advance

**Comment by georgewsinger** [comment]
Any progress on this? Has a single person confirmed that the Vive Pro can run in Linux?

**Comment by rbaumg** [comment]
I am also getting the 108 error, and the Vive Pro always shows as disconnected in the Xorg logs. I have yet to get the Vive Pro to display anything under Linux. 

I'm using the official Nvidia drivers 410.93 (on Bionic as well) for testing this - looking at the release notes for Xorg and the comment about requiring 1.20 or higher for AMD in the README leads me to think it might be related to Xorg 1.19, which is packaged with Bionic.

I plan to upgrade to Cosmic today and will post here if I have better luck.

**Comment by rbaumg** [comment]
Sorry for not getting back to this for a while. I'll try to make up for it with some details; if you want the short version scroll to the end.

I can confirm that the Vive Pro is working with SteamVR on kernel 4.18 + Xorg 1.20.1 (included in Ubuntu Cosmic 18.10) and the NVIDIA binary drivers 410.93. It is actually working quite well with my HTC Vive Pro on an Ubuntu 18.10 test HVM (using PCI passthrough for GPU and USB), all things considered.

I'm not yet sure which version bump allowed it to work (my guess is probably the both), but I know that X.org completed VR HMD support as part of their 1.20 release cycle and the kernel had some changes in v4.15 so I imagine you likely need to upgrade both the kernel _and_ Xorg.

The Linux kernel had some important changes to support the Vive Pro, but Ubuntu maintains their own kernel, and do not always backport patches like they should. As a result, the kernel that ships with Xenial is missing important v4.15 kernel patches. You can check this for yourself by checking their Git log (https://git.launchpad.net/~ubuntu-kernel/ubuntu/+source/linux/+git/xenial/log/).

The kernel commits I found that appear to be relevant here:
- acb1d8e: drm/edid: quirk HTC vive headset as non-desktop. [v2]
- 66660d4: drm: add connector info/property for non-desktop displays [v2]
- c353bfc: Merge tag 'drm-for-v4.15-part2' of git://people.freedesktop.org/~airl…

I was not able to narrow down the specific commits for X.org (please share if you know them), but there is mention of VR HMD support on Phoronix blog:
- https://www.phoronix.com/scan.php?page=news_item&px=Xorg-Server-Entering-2019

I will post back if I happen to learn anything more; for now I'm just starting to test everything. A couple things worth mentioning right away:
- Direct Rendering mode does not work; neither does SteamVR restart (just quits).
- The Headset Firmware updater fails every time. All other firmware installers work fine.
- Framerate is lower than expected (but this could be just this particular testing setup).

TL;DR - SteamVR will not work with Bionic or even Xenial; you will need to upgrade to Cosmic or later.

**Comment by kisak-valve** [comment]
It should be noted that Ubuntu 18.04 currently has xorg 1.20.1 in the hwe stack with the `xserver-xorg-hwe-18.04` package pulling it in. Be careful if using it with a PPA with an xorg ddx because that could be built against the older X server.

**Comment by rbaumg** [comment]
@kisak-valve Thanks for clarifying. Indeed '''xserver-xorg-hwe-18.04''' should be enough to get the Vive to work on Bionic, although I'm not able to test that at the moment. 

I'm not sure if the Vive will work with open-source DDX packages - I've only had success with the binary NVidia drivers (I don't have AMD cards to test with). They will need to be rebuilt (rerun the installer is probably easiest) after updating Xorg and/or kernel version.

**Comment by rwbaumg** [comment]
Just another quick update; Direct Rendering Mode (DRM) is working on Cosmic :+1: . I'm seeing a strange glitch that makes it appear as though Xorg has crashed just as the headset is connecting, but once the HMD comes up the desktop comes back and no errors in the system logs. I haven't really gone through the Steam logs yet, so I may post another update here if I find anything that might be relevant to this ticket - otherwise this issue is resolved for me.

In my opinion the main issue here is poor documentation with regards to SteamVR and HMD requirements, but there are definitely some other issues I would like to pursue in other tickets over the next few weeks. I'm also starting to wonder what might be required to get headsets like the Vive Pro Eye to work under Linux, or if they are already supported with the current stack.

@georgewsinger: Have you had any success getting your Vive Pro to work yet?

**Comment by Casuallynoted** [comment]
Attempting now with my Vive Pro, but it doesn't see the base stations at all. Can confirm it's working on the same computer when in Windows. Controllers do pair and the headset appears in the list but the steamvr menu doesn't come up.

**Comment by Casuallynoted** [comment]
>Installed 18.10 Ubuntu (Kernel 4.18+, XOrg 1.20.1, Latest nVidia driver, 1.3.22 SteamVR beta), but still getting 108 error with Vive Pro (Lighthouse 2.0 system): ...

@rbt2008 Where are logs located? Thinking they may be able to help me figure out what’s going on with my Pro as well.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by kisak-valve** [comment]
Might as well try adding `KERNEL=="hidraw*", SUBSYSTEM=="hidraw", ATTRS{idVendor}=="28de", ATTRS{idProduct}=="2300", TAG+="uaccess"` to /lib/udev/rules.d/60-HTC-Vive-perms.rules and reload the rules then re-plug the vive (or just reboot) and check if it makes a difference.

**Comment by neopostmodern** [comment]
`ATTRS{idProduct}=="2300"` solved it for me; could that please be added to the README?

**Comment by kisak-valve** [comment]
The VR udev rule in the Valve-provided Steam package has had usb ids for the Vive Pro since Steam 1.0.0.55. The README.md has been updated to point users to the [steam-devices repo](https://github.com/ValveSoftware/steam-devices/blob/master/60-steam-vr.rules) which should get updated when more devices are added to udev rule in the Steam package.

If you're using a distro-provided Steam package that is missing 28de:2300, their udev rule for HMDs will need to be updated by distro's the package maintainer for Steam or add the steam-devices variant linked above.
